File: bbq-core/src/trader/position.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};

use super::quot::QuotBar;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields, default, rename_all = "snake_case")]
pub struct Position {
    pub position_id: String,

    // 股票名称
    pub name: String,
    // 股票代码
    pub code: String,
    // 首次建仓时间
    pub time: Option<NaiveDateTime>,

    // 持仓量
    pub volume: u32,
    // 可用持仓量
    pub volume_available: u32,
    // 可用持仓量
    pub volume_frozen: u32,
    // 持仓费用
    pub fee: f64,
    // 平均持仓价
    pub price: f64,
    // 最新价
    pub now_price: f64,
    // 最高价
    pub max_price: f64,
    // 最低价
    pub min_price: f64,
    // 盈利比例
    pub profit_rate: f64,
    // 最大盈利比例
    pub max_profit_rate: f64,
    // 最小盈利比例
    pub min_profit_rate: f64,

    // 盈利
    pub profit: f64,
    // 最大盈利
    pub max_profit: f64,
    // 最小盈利
    pub min_profit: f64,

    // 最大盈利时间
    pub max_profit_time: Option<NaiveDateTime>,
    // 最小盈利时间
    pub min_profit_time: Option<NaiveDateTime>,
}

impl Position {
    pub fn on_update_quot(&mut self, quot_bar: &QuotBar) {
        self.now_price = quot_bar.close;
        if self.max_price < self.now_price {
            self.max_price = self.now_price;
        }
        if self.min_price > self.now_price {
            self.min_price = self.now_price
        }
        self.profit = (self.now_price - self.price) * self.volume as f64 - self.fee;
        self.profit_rate = self.profit / (self.price * self.volume as f64 + self.fee);
        let t = NaiveDateTime::parse_from_str(quot_bar.quot.time.as_str(), "%Y-%m-%d %H:%M:%S");
        if self.profit > self.max_profit {
            self.max_profit = self.profit;
            if let Ok(time) = &t {
                self.max_profit_time = Some(time.clone());
            }

            self.max_profit_rate = self.profit_rate;
        }

        if self.profit < self.min_profit {
            self.min_profit = self.profit;
            if let Ok(time) = &t {
                self.min_profit_time = Some(time.clone());
            }
            self.min_profit_rate = self.profit_rate;
        }
    }
}

File: bbq-strategy-py/qstrategy/position.py
from datetime import datetime
import sys


class Position:
    def __init__(self, js):
        self.position_id = js['position_id']
        self.name = js['name']
        self.code = js['code']
        self.volume = js['volume']
        self.volume_available = js['volume_available']
        self.fee = js['fee']
        self.price = js['price']
        self.profit_rate = js['profit_rate']
        self.max_profit_rate = js['max_profit_rate']
        self.min_profit_rate = js['min_profit_rate']
        self.profit = js['profit']
        self.max_profit = js['max_profit']
        self.min_profit = js['min_profit']
        self.now_price = js['now_price']
        self.max_price = js['max_price']
        self.min_price = js['min_price']
        self.max_profit_time = None
        self.min_profit_time = None
        if js['max_profit_time'] is not None and isinstance(js['max_profit_time'], str):
            s = js['max_profit_time'][:-7]
            self.max_profit_time = js['max_profit_time']
            self.max_profit_time = datetime.strptime(s, '%Y-%m-%dT%H:%M:%S')

        if js['min_profit_time'] is not None and isinstance(js['min_profit_time'], str):
            s = js['min_profit_time'][:-7]
            self.min_profit_time = js['min_profit_time']
            self.min_profit_time = datetime.strptime(s, '%Y-%m-%dT%H:%M:%S')

    def to_dict(self):
        d = {}
        d['position_id'] = self.position_id
        d['name'] = self.name
        d['code'] = self.code

        d['volume'] = self.volume
        d['volume_available'] = self.volume_available
        d['fee'] = self.fee
        d['price'] = self.price
        d['profit_rate'] = self.profit_rate
        d['max_profit_rate'] = self.max_profit_rate

        d['min_profit_rate'] = self.min_profit_rate
        d['profit'] = self.profit
        d['max_profit'] = self.max_profit
        d['min_profit'] = self.min_profit
        d['now_price'] = self.now_price
        d['max_price'] = self.max_price
        d['min_price'] = self.min_price

        d['max_profit_time'] = None if self.max_profit_time is None else self.max_profit_time.strftime(
            "%Y-%m-%dT%H:%M:%S.%f")
        d['min_profit_time'] = None if self.min_profit_time is None else self.min_profit_time.strftime(
            "%Y-%m-%dT%H:%M:%S.%f")

        return d
